src/components/GroupForm/BulkUserManager/BulkUserManager.js

```diff
@@ -0,0 +1,139 @@
+import i18n from '@dhis2/d2-i18n'
+import { DataTableToolbar, Pagination, SegmentedControl } from '@dhis2/ui'
+import PropTypes from 'prop-types'
+import React, { useState, useEffect } from 'react'
+import styles from './BulkUserManager.module.css'
+import { usePendingChanges } from './hooks/usePendingChanges'
+import { useUsers } from './hooks/useUsers'
+import PendingChanges from './PendingChanges'
+import TopBar from './TopBar'
+import UserTable from './UserTable'
+
+const BulkUserManager = ({ groupId, onChange }) => {
+    const [mode, setMode] = useState(groupId ? 'MEMBERS' : 'NON_MEMBERS')
+    const {
+        loading,
+        error,
+        users,
+        pager,
+        setPage,
+        selected,
+        toggleSelected,
+        toggleAllSelected,
+        filter,
+        setFilter,
+    } = useUsers({
+        groupId,
+        mode,
+    })
+    const pendingChanges = usePendingChanges()
+
+    useEffect(() => {
+        onChange(
+            pendingChanges.map(({ action, userId }) => ({ action, userId }))
+        )
+    }, [pendingChanges])
```

Comment:
`onChange` was not added as a `useEffect` dependency here as React Final Form provides fields with a new `onChange` when their value is updated, which would trigger an infinite render loop in this case.